src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo) {
+      const op = await this.getOperation(opInfo);
+      if (op.previousOperationHash) {
+        return op.previousOperationHash;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Return the first version of a DID document given a possibly later version.
+   * A simple recursive implementation using prev; not very efficient but should
+   * not matter since this method is not hit for any externally exposed DID
+   * operations.
+   */
+  public async first (versionId: VersionId): Promise<VersionId | null> {
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo === undefined) {
+      return null;
+    }
+
+    if (this.isInitialVersion(opInfo)) {
+      return versionId;
+    }
+
+    const prevVersionId = await this.prev(versionId);
+    if (prevVersionId === null) {
+      return null;
+    }
+
+    return this.first(prevVersionId);
+  }
+
+  /**
+   * Return the next version of a DID document if it exists or null, otherwise.
+   */
+  public next (versionId: VersionId): VersionId | null {
+    const nextVersionId = this.nextVersion.get(versionId);
+    if (nextVersionId === undefined) {
+      return null;
+    } else {
+      return nextVersionId;
+    }
+  }
+
+  /**
+   * Return the latest (most recent) version of a DID document. Return null if
+   * the version is unknown.
+   */
+  public last (versionId: VersionId): VersionId | null {
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo === undefined) {
+      return null;
+    }
+
+    const nextVersionId = this.next(versionId);
+    if (nextVersionId === null) {
+      return versionId;
+    } else {
+      return this.last(nextVersionId);
+    }
+  }
+
+  /**
+   * Get the last processed transaction.
+   * TODO: fix this after discussing the intended semantics.
+   */
+  public get lastProcessedTransaction (): Transaction | undefined {
+    return undefined;
+  }
+
+  /**
+   * Get a cryptographic hash of the write operation. Currently, uses
+   * SHA256 to get hashes (TODO: Fix it to be consistent DID generation)
+   */
+  private static getHash (operation: WriteOperation): OperationHash {
+    const sha256HashCode = 18;
+    const multihash = Multihash.hash(operation.operationBuffer, sha256HashCode);
+    const multihashBase58 = Base58.encode(multihash);
+    return multihashBase58;
+  }
+
+  /**
+   * Apply state changes for operations that have a previous version (update, delete, recover)
+   */
+  private applyOpWithPrev (opHash: OperationHash, opInfo: OperationInfo, version: VersionId): void {
+    // We might already know of an update to this version. If so, we retain
+    // the older of previously known update and the current one
+    const prevUpdateHash = this.nextVersion.get(version);
+    if (prevUpdateHash !== undefined) {
+      const prevUpdateInfo = this.opHashToInfo.get(prevUpdateHash) as OperationInfo;
```

Comment:
Should this be OperationTimestamp instead of OperationInfo? Since the code make a point of separating them out.

Reply:
This issue should be auto-resolved given the fix for a previous comment (making timestamp a field/property of OperationInfo) - please check.